### PR TITLE
feat(lang): Add proposal for custom base image

### PR DESCRIPTION
**Pull request by gaocegege**
Preview: https://github.com/gaocegege/envd/blob/base/docs/proposals/20220707-base-image.md

Signed-off-by: Ce Gao <cegao@tensorchord.ai>

### CODE REVIEW

File: docs/proposals/20220707-base-image.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    # contains them.
+    base(os="ubuntu", version="20.04")
+```
+
+#### Default image for serving or other purposes
+
+```python
+def build():
+    # We consider it as an image not for development if there is
+    # no image specified in `base` func.
```
Comment (kemingy) [inline docs/proposals/20220707-base-image.md:103]: "no image specified" or "it disables ssh"?

File: docs/proposals/20220707-base-image.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+```python
+def build():
+    base(image="docker.io/tensorchord/custom-image")
+```
+
+New argument `image` is introduced in `base` to support customization. There are some requirements for the image.
+
+- Should be Ubuntu-based image
+- Should have `pip` installed
```
Comment (kemingy) [inline docs/proposals/20220707-base-image.md:63]: Do we need to have a checker or just document this?

Comment (gaocegege) [reply docs/proposals/20220707-base-image.md:63]: I think we should have a check in envd. envd will return an error if there is no pip installed.

**Comment by VoVAllen** [comment]
Some ideas:

We can provide a way for user override `base` function in their own environment, with some methods marked as internal (starting with `_`) now. 

For example
```
def base(os, language):
    _base_img= Image("ubuntu:20.04")
    # Execute command over base
    install_htop = _base_img.run("sudo apt install htop")
    
    # as llb merge, create a new state as base the base image
    _base.image = _base.merge([install_htop, ...])

    # Later step will use _base.image as the base

```

**Comment by gaocegege** [comment]
@VoVAllen 

I am wondering how to support different bases on different targets. For example, I want to use `nvidia/cuda:runtime` for serving, and `nvidia/cuda:devel` for development.

**Comment by VoVAllen** [comment]
Also there are some ambiguity in the current API design. 
`config.ssh(mode='disable')` doesn't sounds good. 

Other options:
- Introduce `install.dev_packages(..)`
- Add `dev` as the argument to the base function, for example, `base(os="ubuntu20.04", dev=default_dev_packages())`

**Comment by gaocegege** [comment]
> Also there are some ambiguity in the current API design. `config.ssh(mode='disable')` doesn't sounds good.
> 
> Other options:
> 
> * Introduce `install.dev_packages(..)`
> * Add `dev` as the argument to the base function, for example, `base(os="ubuntu20.04", dev=default_dev_packages())`

Yep, the main challenge here is how to let us know if we should install the dev dependencies. The approach you proposed LGTM.

**Comment by VoVAllen** [comment]
> @VoVAllen
> 
> I am wondering how to support different bases on different targets. For example, I want to use `nvidia/cuda:runtime` for serving, and `nvidia/cuda:devel` for development.

Probably we want extend the `base` function.

```python
def base(os, requirements):
    # requirements is a limited set maintained by envd, only certain package will be supported
    # Need clear line between requirements here and other `install` packages
    ...

base("ubuntu20.04", default_dev()+cuda("11.1", cudnn="8")) # default_dev()+cuda() is just list concatenation 
# Then envd can use a cuda base image + install other package by apt install 
# user can override base function by their own. The implementation can be flexible
```

Then for serving and dev
```
def build():
   base("ubuntu20.04", default_dev()+cuda("11.1", cudnn="8"))

def serving():
   base("ubuntu20.04", default_dev()+cuda("11.1", cudnn="8"))
```

**Comment by gaocegege** [comment]
Cannot understand the design, could you please give me an example of it.

The deployment image uses nvidia/cuda:devel, and the serving image uses nvidia/cuda:runtime.

**Comment by VoVAllen** [comment]
```
def dev():
   base("ubuntu20.04", default_dev()+cuda("11.1", cudnn="8", version="devel"))

def serving():
   base("ubuntu20.04", cuda("11.1", cudnn="8", version="runtime"))
```

Then envd can handle the logic inside `base` function
For `dev` function, envd will use nvidia base image if found `ubuntu20.04+CUDA dependency`, and install other `default_dev()` requirements by `apt install` or other command. Or base can also contain certain cache logic, to use other pre-built images including all requirements described here. The implementation can be flexible.

**Comment by VoVAllen** [comment]
Also user may not want to include `config.envd` when building serving image

**Comment by gaocegege** [comment]
> ```
> def dev():
>    base("ubuntu20.04", default_dev()+cuda("11.1", cudnn="8", version="devel"))
> 
> def serving():
>    base("ubuntu20.04", cuda("11.1", cudnn="8", version="runtime"))
> ```
> 
> Then envd can handle the logic inside `base` function For `dev` function, envd will use nvidia base image if found `ubuntu20.04+CUDA dependency`, and install other `default_dev()` requirements by `apt install` or other command. Or base can also contain certain cache logic, to use other pre-built images including all requirements described here. The implementation can be flexible.

Users want to specify the custom image instead of using `cuda` function here. For example, users want to use `my-own-org/cuda:devel` in envd. I proposed `base(image="")`. I am wondering how to support this case with your suggestion.

**Comment by aseaday** [comment]
As for this and https://github.com/tensorchord/envd/issues/581: 
Maybe we should use a self custome python installer for those image may do not have apt or package manager?
@gaocegege WDYT

**Comment by gaocegege** [comment]
Yep I think so.

> As for this and #581: Maybe we should use a self custome python installer for those image may do not have apt or package manager? @gaocegege WDYT

**Comment by VoVAllen** [comment]
The problem I'm concerning is the composition for multiple dependency. For example, It's not efficient to construct over all available cuda image to satisfy user requirements. Therefore I'm thinking of dynamically construct base image. If the user requirement is only specify one image, your proposal is totally fine

**Review by gaocegege — COMMENTED** [review]
I am merging it now. But feel free to leave review comments here.

**Pull request merged**